### ISSUE TITLE
[Bug]: elgg_get_friendly_title() might split multi-byte characters due to 100 character limit

### ISSUE THREAD

**Issue by iionly** [Elgg/Elgg#14577]
### Elgg version

4.2 and newer

### Source of your Elgg installation

Elgg.org website

### Describe the bug

Elgg 4.2 introduced a 100-character limit in elgg_get_friendly_title() to fix https://github.com/Elgg/Elgg/issues/13228.

But multi-byte-charecters (2, 3, 4 bytes) that happen to be in the title string at around position 100 might be partly cut off. With incomplete characters at the end of the "friendly" url Apache throws a "AH10244: invalid URI path" error and not serving any underlying file behind that URL.

Use-case exampĺe: iZAP videos plugin off-server videos from Youtube. Youtube allows for example emoticons and also Japanese/Chinese characters in titles of videos. Title length limit is at least >100. The Youtube api serves the title and putting this title through elgg_get_friendly_title() has the risk of returning an invalid URL to serve thumbnail and video if there happen to be multi-byte characters at the wrong position.

### Steps to reproduce

1.
2.
3.
...

**Comment by jeabakker** [comment]
do you have any examples of title that are cut of at the wrong place?

**Comment by iionly** [comment]
e.g.
Original title: "Morihei Ueshiba O Sensei - Rare Aikido Demonstration (1957) 合気道植芝 盛平" (https://www.youtube.com/watch?v=30Sa0PLquFg)

Returned by elgg_get_friendly_title(): "morihei-ueshiba-o-sensei-rare-aikido-demonstration-1957-%E5%90%88%E6%B0%97%E9%81%93%E6%A4%8D%E8%8A%9"

**Comment by jeabakker** [comment]
So far i'm unable to reproduce the issue :(
- which version of apache are you using
- do you have the php module mbstring enabled or not
- where do you see the error (won't the page load, in the apache error log, in the php error log)

**Comment by iionly** [comment]
Apache 2.4.51
mbstring installed and working

The error is from Apache and logged in the Apache error log.

Interesting to note is that if I add a blog or file, the friendly url is different from the return value of elgg_get_friendly_url():

* blog/file: morihei-ueshiba-o-sensei-rare-aikido-demonstration-1957-%25E5%2590%2588%25E6%25B0%2597%25E9%2581%2593%25E6%25A4%258D%25E8%258A%259 (more than 100 characters)
* izap_videos using elgg_get_friendly_title() unchanged: morihei-ueshiba-o-sensei-rare-aikido-demonstration-1957-%E5%90%88%E6%B0%97%E9%81%93%E6%A4%8D%E8%8A%9 (100 characters)
* izap_videos using elgg_get_friendly_title() without cutoff: morihei-ueshiba-o-sensei-rare-aikido-demonstration-1957-%25E5%2590%2588%25E6%25B0%2597%25E9%2581%2593%25E6%25A4%258D%25E8%258A%259 (again more than 100 characters)

The error occurs only in the second case. Could it be that elgg_substr() (mb_substr() respectively) with the string already processed by html_entity_decode() causes the problem if there's a max limit additionally?

What's different with blog and files? When not using elgg_get_friendly_title() what's used instead and is the title even shortened?

**Comment by jeabakker** [comment]
It's the url generator which does some additional encoding
For blog/file the url generator gets the parameter title as `morihei-ueshiba-o-sensei-rare-aikido-demonstration-1957-%E5%90%88%E6%B0%97%E9%81%93%E6%A4%8D%E8%8A%9`

this will then be encoded again into `morihei-ueshiba-o-sensei-rare-aikido-demonstration-1957-%25E5%2590%2588%25E6%25B0%2597%25E9%2581%2593%25E6%25A4%258D%25E8%258A%259`

the change is that all `%` in the 'original' string are converted to `%25`

So could it be that you don't use the route system to generate the URL?

**Comment by jeabakker** [comment]
@iionly could you test something for us?

Could you make a change to the `elgg_get_friendly_title()` function 
https://github.com/Elgg/Elgg/blob/f3a773872b63fa7e1fb7bae0366ab63edf6dbd9e/engine/lib/output.php#L192-L207

replace the last few lines with
```php
// titles are often stored HTML encoded 
$title = html_entity_decode($title ?? '', ENT_QUOTES, 'UTF-8'); 

// limit length to prevent issues with too long URLS (Request-URI Too Large) 
$title = elgg_substr($title, 0, 100);

return \Elgg\Translit::urlize($title); 
```

does that solve your issue?

**Comment by iionly** [comment]
YES!

This way around it works.

(Would you mean usage of elgg_generate_url() with "use of route system"? Actually used in one occation but still the title segment runs through  elgg_get_friendly_title() first. Maybe not necessary then but I wasn't aware. I think I can also change the usage in the other locations then to elgg_generate_url()).

**Comment by jeabakker** [comment]
> (Would you mean usage of elgg_generate_url() with "use of route system"? Actually used in one occation but still the title segment runs through elgg_get_friendly_title() first. Maybe not necessary then but I wasn't aware. I think I can also change the usage in the other locations then to elgg_generate_url()).

yes thats what i meant

**Comment by jeabakker** [comment]
fixed in #14586
